lightning-invoice/src/utils.rs

```diff
@@ -267,29 +268,47 @@ where
 	// the hints across our real nodes we add one hint from each in turn until no node has any hints
 	// left (if one node has more hints than any other, these will accumulate at the end of the
 	// vector).
-	let mut invoice_hints: Vec<RouteHint> = Vec::new();
-	let mut hint_idx = 0;
+	const MAX_HINTS: usize = 3;
 
-	loop {
-		let mut remaining_hints = false;
+    zip_nested_vectors(phantom_hints, MAX_HINTS)
 
-		for hints in phantom_hints.iter() {
-			if invoice_hints.len() == 3 {
-				return invoice_hints
-			}
-
-			if hint_idx < hints.len() {
-				invoice_hints.push(hints[hint_idx].clone());
-				remaining_hints = true
-			}
-		}
-
-		if !remaining_hints {
-			return invoice_hints
-		}
+}
 
-		hint_idx +=1;
-	}
+// Zip multiple nested vectors
+// Similar to the zip method on a vector, it combines the vectors by index - zero index comes first
+// then second, etc.
+// The difference is that this function handles many or just one vector, whereas the builtin method
+// works on just two vectors.
```

Comment:
There's a larger difference, though. `zip` returns `None` when either vector is exhausted. Whereas this iterator will continue to return elements from the non-exhausted vectors.

Reply:
I don't know that that is a true statement about this function.  Because of the break at `result_size` (in this calling function 3), when `next` is called more than 3 times, `None` is returned.  For instance with the first test case of this function, only three items are "collected" which would mean that it hit `None`.

If I misunderstand your point, let me know.

Reply:
> I don't know that that is a true statement about this function. Because of the break at `result_size` (in this calling function 3), when `next` is called more than 3 times, `None` is returned. For instance with the first test case of this function, only three items are "collected" which would mean that it hit `None`.
> 

Eh, the `result_size` is similar to using `take(result_size)`. If this function returns an iterator, then we should remove the `result_size` logic and just have the caller chain `take`.

But my point stands that if you give, say, two vectors of size 2 and 1, respectively, the iterator will yield three elements while zip would yield a single tuple containing an element from each.